Added a ConfigurableMailer

schmitch opened the pull request:
Hello, this is my Work-In-Progress for a Configurable Mailer.

What I also want to add:

- [x] more unit tests
- [x] feedback what could be changed
- [x] code cleanup
- [x] update the documentation
- [x] Name changes (if any)
- [x] squash the PR's commits


Currently I also found duplicated code, so maybe a cleanup of code would be good.

Also it would be good if the code wouldn't be in a single file.

Also this will keep the old SMTPMailer so compile time injection would still be possible.

Mogztter commented on src/main/scala/play/api/libs/mailer/MailerPlugin.scala:
```diff
@@ -109,6 +112,14 @@ class SMTPMailer @Inject() (smtpConfiguration: SMTPConfiguration) extends Mailer
   }
 }
 
+class ConfigurableMailer @Inject()(smtpConfigurationProvider: Provider[SMTPConfiguration]) extends MailerClient {
```
`SMTPMailer` is configurable *once* so `ConfigurableMailer` may be confusing.
Also we should add `SMTP` in the name to make it clear that it will instanciate an `SMTPMailer`, what about `SMTPReconfigurableMailer` or `SMTPReloadableMailer` ?

Mogztter commented:
Looks good to me, one small inline comment about naming.

> What I also want to add:
 more unit tests
 feedback what could be changed
 code cleanup
 update the documentation
 squash the PR's commits

:+1:

@mkurz What do you think about this addition ?

schmitch commented:
Btw @Mogztter there is one thing that could not be fixed with this.
For that you would need a factory pattern which isn't as easy as it should be with guice.

If you need to have multiple mailer instances with different config. you would need to provide some way of assisted key. (https://github.com/google/guice/wiki/AssistedInject)
However I'm not sure if we should add support for such thing. it would be good to have a example in the documentation. but we shouldn't create a factory by ourself.

mkurz commented:
@Mogztter I think it is a good addition.

Mogztter commented:
> However I'm not sure if we should add support for such thing. it would be good to have a example in the documentation. but we shouldn't create a factory by ourself.

:+1:
Indeed, this is an advanced usage and we should just explain how to do it with Runtime time DI and Compile time DI in the documentation to keep the library as simple as possible.

Mogztter commented:
@schmitch Overall this is really good! nice work, tell me when you're ready, I will do a final review then we're good to go!

schmitch commented:
@Mogztter I renamed the Mailer, I'm still unsure about the docs and unsure about this line:
https://github.com/schmitch/play-mailer/commit/462ccc51d5eab1a7cd2a9cba2efd9fc9c346b8b2#diff-a1708174875ec5c58582d97ce301e053R17

At the current state it wouldn't get a new instance overtime, so when I would change it to a def, it would however wouldn't that break things backwards?

Also I would like to fill out a new issue where I would like to cleanup the code, i.e. split the files, improve some tests, reformat code, remove unnecessary code (MailerConfigurationModule since it is duplicated), is that okay?

Mogztter commented:
> I'm still unsure unsure about this line:
schmitch@462ccc5#diff-a1708174875ec5c58582d97ce301e053R17

> At the current state it wouldn't get a new instance overtime, so when I would change it to a def, it would however wouldn't that break things backwards?

Most (all) of the components in Play! ([I18nComponents](https://github.com/playframework/playframework/blob/master/framework/src/play/src/main/scala/play/api/i18n/Messages.scala#L573), [EvolutionsComponents ](https://github.com/playframework/playframework/blob/master/framework/src/play-jdbc-evolutions/src/main/scala/play/api/db/evolutions/EvolutionsModule.scala#L30), [AllowedHostsComponents ](https://github.com/playframework/playframework/blob/master/framework/src/play-filters-helpers/src/main/scala/play/filters/hosts/AllowedHostsFilter.scala#L79), [DBComponents ](https://github.com/playframework/playframework/blob/master/framework/src/play-jdbc/src/main/scala/play/api/db/DBModule.scala#L45)...) are using `lazy val`.

I think the idea is to provide a *good* default. In our case we are providing a *unique* mailer instance configured with the Play! configuration. Most of the time this is what you want (and this was the default behavior in the previous versions of the library).

If you want to configure your mailer instance differently then you are free to create your own instance (and you don't even need `MailerComponents`):

```scala
class ApplicationComponents(context: Context) extends BuiltInComponentsFromContext(context) {  
  lazy val smtpConfiguration = /* new SMTPConfiguration(...) */
  lazy val applicationController = new controllers.ApplicationScala(new SMTPMailer(smtpConfiguration))
  lazy val assets = new controllers.Assets(httpErrorHandler)
  lazy val router = new Routes(httpErrorHandler, applicationController)
}
```

> I'm still unsure about the docs 

I think we should add the snippet above to the documentation to explain how to use `MailerComponents` and how the `mailerClient` is configured.

> Also I would like to fill out a new issue where I would like to cleanup the code, i.e. split the files, improve some tests, reformat code, remove unnecessary code (MailerConfigurationModule since it is duplicated), is that okay?

Sure, try to open one pull request per change (ie. one pull request to remove unnecessary code, another one to improve tests). This will be easier to review.

> Edit: I'm also unsure about what to test. Currently there is no good way to test if the mailer got a new configuration the second time, 

You can always send two emails and check that `get()` method was called two times ? Or use a custom provider and increment a counter.

> and about the documentation i don't think we should add more here, since the things that should be added a more about Factories which doesn't have anything to do about this behavior.

This is allow good to add documentation to explain the "limitation" of the library. We can explain what the library do and what the library don't (and also why, the reason behind).

schmitch commented:
@Mogztter I think the documentation for compile time DI should be done in a seperate issue (I created #85)
Also I added tests to actually see if the mailer really calls the configuration provider twice.
Currently I'm writing the documentation.

schmitch commented:
@Mogztter do you want to take a look at this pr and check if it is merge ready please?

Mogztter commented:
Sure I will do a code review this Monday. Sorry for the delay it's been a
busy weak.
Le 30 janv. 2016 12:04 PM, "Schmitt Christian" <notifications@github.com> a
écrit :

> @Mogztter <https://github.com/Mogztter> do you want to take a look at
> this pr and check if it is merge ready please?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/playframework/play-mailer/pull/81#issuecomment-177147894>
> .
>

schmitch commented:
Yeah january is mostly a busy month ^^

Mogztter commented on src/main/scala/play/api/libs/mailer/MailerPlugin.scala:
```diff
@@ -109,6 +111,14 @@ class SMTPMailer @Inject() (smtpConfiguration: SMTPConfiguration) extends Mailer
   }
 }
 
+class SMTPReconfigurableMailer @Inject()(smtpConfigurationProvider: Provider[SMTPConfiguration]) extends MailerClient {
```
I think `SMTPDynamicMailer` is a better naming (as opposed to a static mailer client where the configuration is static).

Mogztter commented on user-manual.adoc:
````diff
@@ -46,6 +46,8 @@ play {
 }
 ```
 
+Be aware that the get() method of your CustomSMTPConfigurationProvider will be called multiple times.
````
To draw attention, you should use admonition: http://asciidoctor.org/docs/asciidoc-syntax-quick-reference/#paragraphs

```asciidoc
IMPORTANT: The `get()` method of your CustomSMTPConfigurationProvider will be called each time you send an email.
```

Mogztter replied:
As a consequence, we should recommend to write performant code in `get` method

Mogztter commented on src/test/scala/play/api/libs/mailer/MailerPluginSpec.scala:
```diff
@@ -266,6 +274,12 @@ class MailerPluginSpec extends Specification {
     "provide the Java mocked mailer client" in new WithApplication() {
       app.injector.instanceOf(bind[JMailerClient].qualifiedWith("mock")) must beAnInstanceOf[MockMailer]
     }
+    "call the configuration multiple times" in new WithApplication(applicationWithMockedConfigurationProvider) {
```
"call the configuration each time we send an email"

Mogztter commented on README.adoc:
````diff
@@ -120,6 +120,11 @@ import javax.inject.Inject
 class MyComponent @Inject() (mailerClient: MailerClient)
 ```
 
+Currently the mailerClient will always pull the configuration when calling mailerClient.send(email).
````
I think we should only

schmitch commented:
I updated it, and thanks for your review. Currently I would do a squash before we merge that so as soon as you say that this is merge ready I will rewrite the commit history to a single commit.

Mogztter commented on src/main/scala/play/api/libs/mailer/MailerPlugin.scala:
```diff
@@ -20,7 +20,7 @@ trait MailerComponents {
 class MailerModule extends Module {
   def bindings(environment: Environment, configuration: Configuration) = Seq(
-    bind[MailerClient].to[SMTPMailer],
+    bind[MailerClient].to[SMTPReconfigurableMailer],
```
Don't forget to rename `SMTPReconfigurableMailer` to `SMTPDynamicMailer`

schmitch commented:
@Mogztter could I squash the commit?

Mogztter commented:
Yes sure, then I will proceed with merge. Thanks for your work.

schmitch commented:
Done. Great after that I could tackle the next issues.

Mogztter commented:
LGTM :+1: 
Thanks again ;)
